tests/integration/test_scheduler.py

```diff
@@ -292,3 +294,26 @@ def mock_auto_restart(*a, **k):
 
     assert log_filter(log, level=logging.ERROR, contains=err_msg)
     assert TRACEBACK_MSG in log.text
+
+
+async def test_uuid_unchanged_on_restart(
+    one: Scheduler,
+    scheduler: Callable,
+    start: Callable,
+):
+    """Restart gets UUID from Database:
+
+    See https://github.com/cylc/cylc-flow/issues/5615
+
+    Process:
+       * Create a scheduler then shut it down.
+       * Create a new scheduler for the same workflow and check that it has
+         retrieved the UUID from the Daatabase.
+    """
+    uuid_re = re.compile('CYLC_WORKFLOW_UUID=(.*)')
+    contact_file = Path(one.workflow_run_dir) / '.service/contact'
+    async with start(one):
+        original_uuid = uuid_re.findall(contact_file.read_text())[0]
+    schd = scheduler(one.workflow_name, paused_start=True)
+    async with start(schd):
+        assert schd.uuid_str == original_uuid
```

Comment:
This test passes on master!

The issue is more nuanced, the UUID str *is* getting loaded correctly on master, but *after* the point when the contact file is written.

To detect the old bug you would need to compare the contact file of the second run, maybe something like this:

```python
num_starts = 3
uuids = []
for _ in range(num_starts):
    schd = scheduler(one.workflow_name, paused_start=True)
    async with start(schd):
        uuids.append(schd.uuid_str)
        uuids.append(uuid_re.findall(contact_file.read_text())[0])

assert len(uuids) == num_starts
assert len(set(uuids)) == 1
```

Note there are some other things which use the UUID too, we should probably test those too.

Reply:
!  And I spent ages branch flicking to try and avoid that!